### PR TITLE
rgw: refactor PutObjProcessor stack

### PR DESCRIPTION
**Pull request by cbodley**
this breaks up the PutObjProcessor stack into smaller pieces that are more composable and easier to test. this will an allow asio-style YieldingAioThrottle (not present here) to be swapped in indepedently of the Atomic/Multipart processors, which previously inherited from the Aio processor. there are some notes about the layering and design in https://pad.ceph.com/p/rgw-putobjprocessor

### CODE REVIEW

**Comment by yehudasa** [comment]
@cbodley didn't dive thoroughly, but all in all looks like nice work. We'll need to see how this works with my current refactoring work. Also need to think of how processors for non-rados backends with different semantics (e.g., cloud) will look like, and whether interfaces make sense in that context.

**Comment by mattbenjamin** [comment]
@yehudasa agree on the assessment;  on your point about non-RADOS backends--my intuition has been clear that the general operations should be higher level:  store object;  enumerate objects;  enumeraye buckets;   It feels odd to map these primitives to non-RADOS backends

**Comment by mattbenjamin** [comment]
@cbodley there are aspects of this change I didn't expect, but it looks good

**Comment by cbodley** [comment]
> Also need to think of how processors for non-rados backends with different semantics (e.g., cloud) will look like, and whether interfaces make sense in that context.

i think the DataProcessor interface is a good fit for stream protocols in general, and makes it easy to add compression/encryption/other transformations to the outgoing stream. the StripeProcessor would be useful for a strategy involving multipart uploads. the ObjectProcessor itself is also general enough to make sense for an http backend, and the HeadObjectProcessor could work as well. past that, i don't see much opportunity to reuse the derived Manifest/Atomic/Multipart processors

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -3634,7 +3626,32 @@ void RGWPutObj::execute()
     }
   }
 
-  op_ret = processor->prepare(store, NULL);
+  // create the object processor
+  using namespace rgw::putobj;
+  AioThrottle aio(store->ctx()->_conf->rgw_put_obj_min_window_size);
+  constexpr auto max_processor_size = std::max(sizeof(MultipartObjectProcessor),
+                                               sizeof(AtomicObjectProcessor));
+  ceph::static_ptr<ObjectProcessor, max_processor_size> processor;
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: did we have this before?

Comment (cbodley) [reply src/rgw/rgw_op.cc]: we had been heap-allocating the processor in select_processor(). static_ptr itself merged last year in https://github.com/ceph/ceph/pull/19079

Comment (mattbenjamin) [reply src/rgw/rgw_op.cc]: I remembered the latter, nice to see us using it

File: src/rgw/rgw_file.cc [hunk 2]
```diff
@@ -1324,16 +1324,18 @@ namespace rgw {
     }
 
     filter = processor;
+#if 0
```
[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_file.cc]: i wanted to split the interface changes in processors and compression/encryption filters over several commits. this commit made it possible to compile each commit along the way

**Comment by cbodley** [comment]
http://pulpito.ceph.com/cbodley-2018-10-16_17:32:21-rgw-wip-rgw-putobjprocessor-distro-basic-smithi/

single unrelated multisite failure in test_versioned_object_incremental_sync, and valgrind issues were either in ceph-mon or the beast frontend

**Pull request merged**